src/main/scala/diplomacy/Nodes.scala

```diff
@@ -10,82 +10,398 @@ import freechips.rocketchip.util.HeterogeneousBag
+  *         `index` is current `nodeA` [[OutwardNode.accPO]] size, which represents the how many [[InwardNode]] has connected to `nodeA`
+  *         `node` is [[NodeHandle]] of `NodeB`,
+  *         `binding` is binding type, which is used for deciding port mapping behavior.
+  *       similarly, `nodeB`, will call [[InwardNode.iPush]] after `nodeA` finish its binding.
+  *   2. Port mapping(lazy)
+  *     Port mapping converts [[OutwardNode.accPO]] and [[InwardNode.accPI]] to [[MixedNode.oPortMapping]] and [[MixedNode.iPortMapping]]
+  *     it is implemented with [[MixedNode.resolveStar]], which solves the size of each connection.
+  *   3. Parameter negotiation(lazy)
+  *     After finishing port mapping,
+  *     [[MixedNode.doParams]] and [[MixedNode.uiParams]] are result of parameter negotiation,
+  *   4. Bundle building.(lazy)
+  *     After finishing parameter propagation,
+  *     [[MixedNode.edges]] can be generated by zipping ports and parameters together.
+  *
+  * Diplomacy has some implicit acronym described below:
+  *   Parameters System:
+  *   D[IO], U[IO], E[IO], B[IO] is parameters will be propagated.
+  *   D: Downwards (master -> slave)
+  *   U: Upwards (slave -> master)
+  *   E: Edge contains necessary of nearby nodes connection information.
+  *   B: Bundle is [[chisel3.Data]], which you can attach chisel data type to it.
+  *
+  * Concepts, metaphors, and mnemonics to help with understanding Diplomacy code.
+  *
+  *
+  * # Abstract types
+  *
+  * Diplomacy is a set of abstractions for describing directed, acyclic graphs
+  * where parameters may be negotiated between nodes. These abstractions are
+  * expressed in the form of abstract classes, traits, and type parameters, which
+  * comprise nearly all of the types defined in this file.
+  *
+  * The NodeImp ("node implementation") is the main abstract type that determines
+  * the type parameters of all other abstract types. Defining a concrete
+  * implementation of NodeImp will therefore determine concrete types for all
+  * type parameters. For example, passing in a concrete instance of NodeImp to a
+  * SourceNode will fully determine concrete types for all of a SourceNode's type
+  * parameters.
+  *
+  * Specific applications of Diplomacy are expected to either extend these types
+  * or to specify concrete types for the type parameters. This allows for
+  * creating and associating application-specific node, edge, parameter, and bundle types.
+  *
+  * It is key to understand that the edge parameters can be abstract concept -- it carries any sort of conceptual
+  * information that is useful to pass along the graph edge. For example, the full address map visible from
+  * a given edge including the supported access types for each memory region.
+  * In contrast, the bundle parameter carries information used to parameterize the hardware bundle type (for
+  * example, signal widths).
+  *
+  *
+  *
```

Comment:
```suggestion
  * Anatomy of a Node:
  * 
  * ```
  *                                           |                                       Upwards (aka Source)
  *                                        | inward edge of (parameter) type EI
  * 					  | created from parameters of type UI and DI
  *					  | will result in a Bundle of type BI
  *					  |
  *                                     ^  |  *
  *                                     .  |  *
  *      |------------------------------.--|--*------------------------------------|   
  *      |  |---------------------------.--|--*-----------------------------|      |
  *      |  |                           .  |  *           InwardNodeImp     |      |
  *      |  |                        (mixI)|  *                             |      |
  *      |  |                           .  |  *                             |      |
  *      |  |    Upwards-flowing input  .  |  * Downwards-Flowing input     |    N |
  *      |  |    parameters of type UI  .  |  * parameters of type DI       |    o |
  *      |  |---------------------------.--|--*-----------------------------|    d |          
  *      |                              .  |  *                                  e |
  *      |                              .  v  v                                  I |
  *      |                              .     (mapParamsD)                       m |
  *      |                    (mapParamsU)    +                                  p |
  *      |                              ^  |  +                                    |
  *      |                              :  |  +                                    |
  *      |  |---------------------------:--|--+-----------------------------|      |
  *      |  |                           :  |  +           OutwardNodeImp    |      |
  *      |  |                           :  | (mixO)                         |      |
  *      |  |                           :  |  +                             |      |
  *      |  |  Upwards-flowing output   :  |  + Downwards-Flowing output    |      |
  *      |  |    parameters of type UO  :  |  + parameters of type DO       |      |
  *      |  |---------------------------:--|--+-----------------------------|      |          
  *      |                              :  |  +                                    |
  *      |                              :  |  +                                    |
  *      |------------------------------:--|--+------------------------------------|
  *                                     :  |  v                               
  *					  | outward edge of (parameter) type EO
  *					  | created from parameters of type UO and DO
  *					  | will result in a Bundle of type BO
  *                                      |
  *	                                  v	                                 Downwards (aka Sink)

```

Reply:
I rework on this graph. maybe it's more clear now(I think)